core/coreunix/add.go

```diff
@@ -92,46 +230,88 @@ func AddWrapped(n *core.IpfsNode, r io.Reader, filename string) (string, *merkle
 	return gopath.Join(k.String(), filename), dagnode, nil
 }
 
-func add(n *core.IpfsNode, reader io.Reader) (*merkledag.Node, error) {
-	return importer.BuildDagFromReader(
-		n.DAG,
-		chunk.DefaultSplitter(reader),
-	)
-}
+func (params *Adder) addNode(node *merkledag.Node, path string) error {
+	// patch it into the root
+	if path == "" {
+		key, err := node.Key()
+		if err != nil {
+			return err
+		}
+
+		path = key.Pretty()
+	}
 
-func addNode(n *core.IpfsNode, node *merkledag.Node) error {
-	if err := n.DAG.AddRecursive(node); err != nil { // add the file to the graph + local storage
+	if err := params.editor.InsertNodeAtPath(params.ctx, path, node, newDirNode); err != nil {
 		return err
 	}
-	ctx, cancel := context.WithCancel(n.Context())
-	defer cancel()
-	err := n.Pinning.Pin(ctx, node, true) // ensure we keep it
-	return err
+
+	return outputDagnode(params.out, path, node)
 }
 
-func addFile(n *core.IpfsNode, file files.File) (*merkledag.Node, error) {
-	if file.IsDirectory() {
-		return addDir(n, file)
+// Add the given file while respecting the params.
+func (params *Adder) AddFile(file files.File) (*merkledag.Node, error) {
+	switch {
+	case files.IsHidden(file) && !params.Hidden:
+		log.Debugf("%s is hidden, skipping", file.FileName())
+		return nil, &hiddenFileError{file.FileName()}
+	case file.IsDirectory():
+		return params.addDir(file)
+	}
+
+	// case for symlink
+	if s, ok := file.(*files.Symlink); ok {
+		sdata, err := unixfs.SymlinkData(s.Target)
+		if err != nil {
+			return nil, err
+		}
+
+		dagnode := &merkledag.Node{Data: sdata}
+		_, err = params.node.DAG.Add(dagnode)
+		if err != nil {
+			return nil, err
+		}
+
+		err = params.addNode(dagnode, s.FileName())
+		return dagnode, err
+	}
+
+	// case for regular file
+	// if the progress flag was specified, wrap the file so that we can send
+	// progress updates to the client (over the output channel)
+	var reader io.Reader = file
+	if params.Progress {
+		reader = &progressReader{file: file, out: params.out}
 	}
-	return add(n, file)
-}
 
-func addDir(n *core.IpfsNode, dir files.File) (*merkledag.Node, error) {
+	dagnode, err := params.add(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	// patch it into the root
+	log.Infof("adding file: %s", file.FileName())
+	err = params.addNode(dagnode, file.FileName())
+	return dagnode, err
+}
 
-	tree := &merkledag.Node{Data: unixfs.FolderPBData()}
+func (params *Adder) addDir(file files.File) (*merkledag.Node, error) {
```

Comment:
even though the type is `files.File` we should probably stick to naming the variable  `dir` since thats what it is here